Epic: IAM Hardening — Security, Reliability & SCIM Compliance Fixes

## Context

PR #34 (IAM — Keycloak User CRUD, SCIM v2 & OPA Authorization) was merged into `rc` with a comprehensive code review identifying 23 findings across security, reliability, and spec compliance. This Epic addresses the critical and high-severity items that need hardening before promoting to `main`.

**Source**: [PR #34 Code Review](https://github.com/younjinjeong/microfoundry/pull/34#issuecomment-3914621347)

---

## Phase 1: Critical Security Fixes

### 1.1 — Rego policy: restrict SCIM when auth disabled
**File**: `pkg/auth/policies/authz.rego`  
**Issue**: `allow if { input.user == null }` permits unauthenticated SCIM access  
**Fix**: Exclude SCIM/settings/clusters resources from the auth-disabled allow-all rule:
```rego
allow if {
    input.user == null
    not input.resource in {"scim", "settings", "clusters"}
}
```

### 1.2 — OPA middleware: log OrgStore errors
**File**: `pkg/auth/opa_middleware.go:103`  
**Issue**: `members, _ := orgStore.ListMembers(...)` silently swallows errors  
**Fix**: Log the error for observability (don't deny — orgStore failure shouldn't block reads)

### 1.3 — SCIM JSON encoding: handle errors
**File**: `pkg/admin/scim_handlers.go:16-18`  
**Issue**: `json.NewEncoder(w).Encode(v)` error discarded  
**Fix**: Log encoding errors (can't change HTTP status after WriteHeader)

---

## Phase 2: High Reliability Fixes

### 2.1 — IAM handlers: log silent failures
**File**: `pkg/admin/iam_handlers.go:47,63,93,106`  
**Issue**: `GetUserOrgs`, `ListMembers`, `GetUserRoles`, `GetRealmRoles` errors swallowed  
**Fix**: Log errors and pass `Error` field to templates for display

### 2.2 — Password reset: handle error
**File**: `pkg/admin/iam_handlers.go:184`  
**Issue**: `_ = s.keycloakAdmin.ResetPassword(...)` ignored  
**Fix**: Return error to user if password reset fails after user creation

### 2.3 — OrgStore nil guard
**File**: `pkg/admin/iam_handlers.go:46`, `pkg/admin/org_handlers.go:38`  
**Issue**: `s.orgStore()` returns nil when auth not configured, calling methods panics  
**Fix**: Add nil check before using orgStore in all tab handlers

### 2.4 — SCIM user ID validation
**File**: `pkg/admin/scim_handlers.go:130,148,181,235`  
**Issue**: `r.PathValue("id")` used without format validation  
**Fix**: Validate UUID format before passing to Keycloak API

---

## Phase 3: SCIM Spec Compliance (Medium)

### 3.1 — SCIM count upper bound
**File**: `pkg/admin/scim_handlers.go:44`  
**Fix**: Cap `count` at 100; return error for values > 200

### 3.2 — ServiceProviderConfig accuracy
**File**: `pkg/admin/scim_handlers.go:244`  
**Fix**: Update discovery response to accurately reflect supported PATCH paths and filter operators

### 3.3 — OPA UpdatePolicy atomicity
**File**: `pkg/auth/opa.go:132-138`  
**Fix**: Only update modules map after successful compile (copy-on-write pattern)

### 3.4 — SCIM filter parsing improvement
**File**: `pkg/auth/scim.go:146-170`  
**Fix**: Handle quoted values with spaces; return specific error for unsupported compound filters

---

## Acceptance Criteria

- [ ] `go build ./...` passes
- [ ] `go vet ./...` passes
- [ ] SCIM endpoints return 401/403 when auth disabled (not 200)
- [ ] OPA middleware logs OrgStore errors
- [ ] No silent `_ =` error swallowing in IAM handlers
- [ ] OrgStore nil check prevents panic when auth disabled
- [ ] SCIM count capped at 100
- [ ] All review findings from PR #34 addressed

## Branch Strategy

```
rc → epic/iam-hardening → PR → rc
```

## 🚀 DevOps Engineer — Deployment, Observability & CI Considerations

### Deployment Risk Assessment

This hardening Epic modifies **security-critical middleware** (OPA, auth). Here's my deployment strategy recommendation:

---

### Rolling Update Safety

The changes in this Epic are **backwards-compatible** — they tighten security rules and add error handling without changing API contracts. Safe for rolling update. However:

**Rego policy change (§1.1)** will immediately start **rejecting unauthenticated SCIM requests** in deployments where auth is disabled. If any external system is currently using SCIM without auth, this will break their integration.

**Pre-deployment checklist:**
- [ ] Confirm no SCIM clients exist in non-auth environments
- [ ] If they exist, coordinate SCIM token/auth setup before deploying
- [ ] Test the updated Rego policy in a staging cluster first

---

### Observability Improvements

The error logging additions (§1.2, §1.3, §2.1) should follow a consistent pattern that integrates with our Loki stack:

**Recommended log format** (structured, parseable by Loki):
```go
log.Printf("[WARN] iam: orgStore.ListMembers failed org=%s err=%v", orgID, err)
log.Printf("[ERROR] scim: json encode failed path=%s err=%v", r.URL.Path, err)
log.Printf("[WARN] iam: keycloak.GetUserRoles failed user=%s err=%v", userID, err)
```

**Log label conventions for Loki queries:**
```
{app="mf-admin"} |= "iam:" | logfmt | level="WARN"
{app="mf-admin"} |= "scim:" | logfmt | level="ERROR"
```

This makes it easy to create Grafana dashboards for:
- SCIM error rates
- IAM operation failures
- OPA deny rates (from audit log)

---

### Audit Log Persistence

The current 1000-entry ring buffer loses all data on restart. For production, I recommend:

**Phase 1 (this Epic):** Add a structured log output alongside the ring buffer:
```go
func (al *AuditLog) Record(entry AuditEntry) {
    // Existing ring buffer logic
    al.mu.Lock()
    // ...
    al.mu.Unlock()

    // Also emit as structured log (captured by Loki)
    log.Printf("[AUDIT] user=%s action=%s resource=%s path=%s allowed=%t reason=%s",
        entry.UserEmail, entry.Action, entry.Resource, entry.Path, entry.Allowed, entry.Reason)
}
```

**Phase 2 (future Epic):** Persist to a ConfigMap or external store for compliance requirements.

This way, the in-memory buffer serves the Admin UI, and Loki provides long-term retention.

---

### CI Pipeline Additions

For this hardening Epic, I recommend adding these CI checks:

**1. OPA Policy Unit Tests:**
```bash
# Run Rego policy tests
opa test pkg/auth/policies/ -v
```

Create `pkg/auth/policies/authz_test.rego`:
```rego
package microfoundry.authz
import rego.v1

# SCIM must be denied when user is null
test_scim_denied_when_no_user if {
    not allow with input as {"user": null, "resource": "scim", "action": "read"}
}

# Authenticated user can read apps
test_authenticated_user_can_read if {
    allow with input as {"user": {"id": "u1", "roles": [], "org_role": "member"}, "resource": "apps", "action": "read"}
}

# Platform admin can access everything
test_platform_admin_full_access if {
    allow with input as {"user": {"id": "u1", "roles": ["platform-admin"], "org_role": ""}, "resource": "scim", "action": "write"}
}
```

**2. SCIM Endpoint Smoke Tests:**
```bash
# Verify SCIM returns 401/403 when auth is not configured
curl -s -o /dev/null -w "%{http_code}" http://localhost:8080/scim/v2/Users
# Expected: 403 (not 200)
```

---

### Build & Release Notes

When implementing, ensure:
- `go mod tidy` is run after any OPA dependency changes
- `go vet ./...` and `go build ./...` pass cleanly (as they do now)
- Tag the release with hardening notes: `v0.x.y-rc.2` → includes IAM hardening

---

### Branch Strategy

```
rc (current) → epic/iam-hardening → PR → rc → (after validation) → main
```

The Epic should be a single PR with all phases, since the changes are interdependent (e.g., Rego policy change depends on proper error handling being in place to catch any fallout).

--- a/pkg/admin/iam_handlers.go
+++ b/pkg/admin/iam_handlers.go
@@ -2,6 +2,7 @@ package admin
 
 import (
 	"encoding/json"
+	"log"
 	"net/http"
 
 	"github.com/younjinjeong/microfoundry/pkg/auth"
@@ -44,7 +45,16 @@ func (s *Server) renderIAMOrgsTab(w http.ResponseWriter, r *http.Request) {
 
 	if user != nil {
 		store := s.orgStore()
-		orgs, _ := store.GetUserOrgs(r.Context(), user.Email)
+		if store == nil {
+			data["Error"] = "Organization store not available"
+			s.templates.RenderPartial(w, "tab_iam_orgs.html", data)
+			return
+		}
+		orgs, err := store.GetUserOrgs(r.Context(), user.Email)
+		if err != nil {
+			log.Printf("[IAM] failed to get user orgs for %s: %v", user.Email, err)
+			data["Error"] = "Failed to load organizations"
+		}
 
 		selectedOrg := r.URL.Query().Get("org")
 		if selectedOrg == "" && len(orgs) > 0 {
@@ -60,7 +70,10 @@ func (s *Server) renderIAMOrgsTab(w http.ResponseWriter, r *http.Request) {
 			orgDetail, err := store.Get(r.Context(), selectedOrg)
 			if err == nil {
 				data["OrgDetail"] = orgDetail
-				members, _ := store.ListMembers(r.Context(), selectedOrg)
+				members, err := store.ListMembers(r.Context(), selectedOrg)
+				if err != nil {
+					log.Printf("[IAM] failed to list members for org %s: %v", selectedOrg, err)
+				}
 				data["Members"] = members
 
 				userRole := "viewer"
@@ -90,7 +103,11 @@ func (s *Server) renderIAMUsersTab(w http.ResponseWriter, r *http.Request) {
 		} else {
 			// Fetch roles for each user
 			for i := range users {
-				roles, _ := s.keycloakAdmin.GetUserRoles(r.Context(), users[i].ID)
+				roles, err := s.keycloakAdmin.GetUserRoles(r.Context(), users[i].ID)
+				if err != nil {
+					log.Printf("[IAM] failed to get roles for user %s: %v", users[i].ID, err)
+					continue
+				}
 				roleNames := make([]string, 0, len(roles))
 				for _, role := range roles {
 					roleNames = append(roleNames, role.Name)
@@ -103,7 +120,10 @@ func (s *Server) renderIAMUsersTab(w http.ResponseWriter, r *http.Request) {
 		}
 
 		// Get available realm roles for assignment
-		realmRoles, _ := s.keycloakAdmin.GetRealmRoles(r.Context())
+		realmRoles, err := s.keycloakAdmin.GetRealmRoles(r.Context())
+		if err != nil {
+			log.Printf("[IAM] failed to get realm roles: %v", err)
+		}
 		data["RealmRoles"] = realmRoles
 	} else {
 		data["NotConfigured"] = true
@@ -181,7 +201,11 @@ func (s *Server) CreateKeycloakUserHandler(w http.ResponseWriter, r *http.Reques
 	}
 
 	if password != "" {
-		_ = s.keycloakAdmin.ResetPassword(r.Context(), userID, password, false)
+		if err := s.keycloakAdmin.ResetPassword(r.Context(), userID, password, false); err != nil {
+			log.Printf("[IAM] user %s created but password reset failed: %v", userID, err)
+			http.Error(w, "User created but password set failed: "+err.Error(), http.StatusInternalServerError)
+			return
+		}
 	}
 
 	w.Header().Set("HX-Redirect", "/users?tab=users")

--- a/pkg/admin/org_handlers.go
+++ b/pkg/admin/org_handlers.go
@@ -2,6 +2,7 @@ package admin
 
 import (
 	"encoding/json"
+	"log"
 	"net/http"
 
 	"github.com/younjinjeong/microfoundry/pkg/auth"
@@ -36,7 +37,15 @@ func (s *Server) OrgsPageHandler(w http.ResponseWriter, r *http.Request) {
 	case "orgs":
 		ctx := r.Context()
 		store := s.orgStore()
-		orgs, _ := store.GetUserOrgs(ctx, user.Email)
+		if store == nil {
+			content["Error"] = "Organization store not available"
+			break
+		}
+		orgs, err := store.GetUserOrgs(ctx, user.Email)
+		if err != nil {
+			log.Printf("[Orgs] failed to get user orgs for %s: %v", user.Email, err)
+			content["Error"] = "Failed to load organizations"
+		}
 
 		selectedOrg := r.URL.Query().Get("org")
 		if selectedOrg == "" && len(orgs) > 0 {
@@ -51,7 +60,10 @@ func (s *Server) OrgsPageHandler(w http.ResponseWriter, r *http.Request) {
 			orgDetail, err := store.Get(ctx, selectedOrg)
 			if err == nil {
 				content["OrgDetail"] = orgDetail
-				members, _ := store.ListMembers(ctx, selectedOrg)
+				members, err := store.ListMembers(ctx, selectedOrg)
+				if err != nil {
+					log.Printf("[Orgs] failed to list members for org %s: %v", selectedOrg, err)
+				}
 				content["Members"] = members
 
 				userRole := "viewer"
@@ -73,7 +85,11 @@ func (s *Server) OrgsPageHandler(w http.ResponseWriter, r *http.Request) {
 				content["Error"] = err.Error()
 			} else {
 				for i := range users {
-					roles, _ := s.keycloakAdmin.GetUserRoles(r.Context(), users[i].ID)
+					roles, err := s.keycloakAdmin.GetUserRoles(r.Context(), users[i].ID)
+					if err != nil {
+						log.Printf("[Orgs] failed to get roles for user %s: %v", users[i].ID, err)
+						continue
+					}
 					roleNames := make([]string, 0, len(roles))
 					for _, role := range roles {
 						roleNames = append(roleNames, role.Name)
@@ -84,7 +100,10 @@ func (s *Server) OrgsPageHandler(w http.ResponseWriter, r *http.Request) {
 				content["TotalUsers"] = total
 				content["Search"] = search
 			}
-			realmRoles, _ := s.keycloakAdmin.GetRealmRoles(r.Context())
+			realmRoles, err := s.keycloakAdmin.GetRealmRoles(r.Context())
+			if err != nil {
+				log.Printf("[Orgs] failed to get realm roles: %v", err)
+			}
 			content["RealmRoles"] = realmRoles
 		} else {
 			content["NotConfigured"] = true
@@ -292,7 +311,10 @@ func (s *Server) APIGetOrgHandler(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
-	members, _ := store.ListMembers(r.Context(), orgID)
+	members, err := store.ListMembers(r.Context(), orgID)
+	if err != nil {
+		log.Printf("[Orgs] failed to list members for org %s: %v", orgID, err)
+	}
 
 	writeJSON(w, http.StatusOK, map[string]any{
 		"org":     org,

--- a/pkg/admin/scim_handlers.go
+++ b/pkg/admin/scim_handlers.go
@@ -3,17 +3,28 @@ package admin
 import (
 	"encoding/json"
 	"fmt"
+	"log"
 	"net/http"
+	"regexp"
 	"strconv"
 
 	"github.com/younjinjeong/microfoundry/pkg/auth"
 )
 
+// uuidRegex validates UUID format for SCIM path parameters.
+var uuidRegex = regexp.MustCompile(`^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$`)
+
+func isValidUUID(s string) bool {
+	return uuidRegex.MatchString(s)
+}
+
 // writeSCIMJSON writes a SCIM JSON response with the correct content type.
 func writeSCIMJSON(w http.ResponseWriter, status int, v any) {
 	w.Header().Set("Content-Type", "application/scim+json")
 	w.WriteHeader(status)
-	json.NewEncoder(w).Encode(v)
+	if err := json.NewEncoder(w).Encode(v); err != nil {
+		log.Printf("[SCIM] failed to encode JSON response: %v", err)
+	}
 }
 
 // writeSCIMError sends a SCIM error response.
@@ -46,6 +57,9 @@ func (s *Server) SCIMListUsersHandler(w http.ResponseWriter, r *http.Request) {
 	count, _ := strconv.Atoi(r.URL.Query().Get("count"))
 	if count <= 0 {
 		count = 20
+	}
+	if count > 100 {
+		count = 100
 	}
 
 	filter := r.URL.Query().Get("filter")
@@ -128,6 +142,10 @@ func (s *Server) SCIMGetUserHandler(w http.ResponseWriter, r *http.Request) {
 	}
 
 	userID := r.PathValue("id")
+	if !isValidUUID(userID) {
+		writeSCIMError(w, http.StatusBadRequest, "invalid user ID format")
+		return
+	}
 	user, err := s.keycloakAdmin.GetUser(r.Context(), userID)
 	if err != nil {
 		writeSCIMError(w, http.StatusNotFound, err.Error())
@@ -146,6 +164,10 @@ func (s *Server) SCIMUpdateUserHandler(w http.ResponseWriter, r *http.Request) {
 	}
 
 	userID := r.PathValue("id")
+	if !isValidUUID(userID) {
+		writeSCIMError(w, http.StatusBadRequest, "invalid user ID format")
+		return
+	}
 
 	var scimUser auth.SCIMUser
 	if err := json.NewDecoder(r.Body).Decode(&scimUser); err != nil {
@@ -179,6 +201,10 @@ func (s *Server) SCIMPatchUserHandler(w http.ResponseWriter, r *http.Request) {
 	}
 
 	userID := r.PathValue("id")
+	if !isValidUUID(userID) {
+		writeSCIMError(w, http.StatusBadRequest, "invalid user ID format")
+		return
+	}
 
 	var patch auth.SCIMPatchOp
 	if err := json.NewDecoder(r.Body).Decode(&patch); err != nil {
@@ -233,6 +259,10 @@ func (s *Server) SCIMDeleteUserHandler(w http.ResponseWriter, r *http.Request) {
 	}
 
 	userID := r.PathValue("id")
+	if !isValidUUID(userID) {
+		writeSCIMError(w, http.StatusBadRequest, "invalid user ID format")
+		return
+	}
 	if err := s.keycloakAdmin.DeleteUser(r.Context(), userID); err != nil {
 		writeSCIMError(w, http.StatusInternalServerError, err.Error())
 		return
@@ -255,7 +285,7 @@ func (s *Server) SCIMServiceProviderConfigHandler(w http.ResponseWriter, r *http
 		},
 		"filter": map[string]any{
 			"supported":  true,
-			"maxResults": 200,
+			"maxResults": 100,
 		},
 		"changePassword": map[string]any{
 			"supported": true,

--- a/pkg/auth/opa.go
+++ b/pkg/auth/opa.go
@@ -129,12 +129,29 @@ func (e *OPAEngine) Evaluate(ctx context.Context, input AuthzInput) (AuthzResult
 }
 
 // UpdatePolicy adds or replaces a custom policy module and recompiles.
+// Uses copy-on-write to avoid corrupting modules map if compilation fails.
 func (e *OPAEngine) UpdatePolicy(name, regoText string) error {
 	e.mu.Lock()
 	defer e.mu.Unlock()
 
-	e.modules[name] = regoText
-	return e.compile()
+	// Copy modules to avoid mutation before successful compile
+	updated := make(map[string]string, len(e.modules)+1)
+	for k, v := range e.modules {
+		updated[k] = v
+	}
+	updated[name] = regoText
+
+	compiler, err := ast.CompileModulesWithOpt(updated, ast.CompileOpts{
+		EnablePrintStatements: false,
+	})
+	if err != nil {
+		return fmt.Errorf("compiling rego policies: %w", err)
+	}
+
+	// Swap atomically only after successful compile
+	e.modules = updated
+	e.compiler = compiler
+	return nil
 }
 
 // GetPolicies returns the currently loaded policy names and their source.

--- a/pkg/auth/opa_middleware.go
+++ b/pkg/auth/opa_middleware.go
@@ -1,6 +1,7 @@
 package auth
 
 import (
+	"log"
 	"net/http"
 	"strings"
 	"time"
@@ -100,7 +101,10 @@ func buildAuthzInput(r *http.Request, orgStore *OrgStore) AuthzInput {
 	if user != nil {
 		orgRole := ""
 		if orgStore != nil && input.OrgID != "" {
-			members, _ := orgStore.ListMembers(r.Context(), input.OrgID)
+			members, err := orgStore.ListMembers(r.Context(), input.OrgID)
+			if err != nil {
+				log.Printf("[OPA] failed to list org members for org %s: %v", input.OrgID, err)
+			}
 			for _, m := range members {
 				if m.Email == user.Email {
 					orgRole = m.Role

--- a/pkg/auth/policies/authz.rego
+++ b/pkg/auth/policies/authz.rego
@@ -4,9 +4,11 @@ import rego.v1
 
 default allow := false
 
-# When auth is disabled, user is null — allow everything
+# When auth is disabled, user is null — allow public resources only
+# SCIM, settings, clusters, and user management still require authentication
 allow if {
 	input.user == null
+	not input.resource in {"scim", "settings", "clusters", "users"}
 }
 
 # Platform admins have full access to all resources

--- a/pkg/auth/scim.go
+++ b/pkg/auth/scim.go
@@ -144,10 +144,17 @@ func SCIMUserToKeycloak(scim *SCIMUser) *KeycloakUser {
 }
 
 // ParseSCIMFilter parses a basic SCIM filter expression.
-// Supports: attrName op "value" (e.g., userName eq "john").
+// Supports: attrName op "value" (e.g., userName eq "john doe").
+// Compound filters (and/or) are not supported and will return an error.
 func ParseSCIMFilter(filter string) (attr, op, value string, err error) {
 	if filter == "" {
 		return "", "", "", nil
+	}
+
+	// Reject compound filters — we only support simple expressions
+	lowerFilter := strings.ToLower(filter)
+	if strings.Contains(lowerFilter, " and ") || strings.Contains(lowerFilter, " or ") {
+		return "", "", "", fmt.Errorf("compound filters (and/or) are not supported")
 	}
 
 	parts := strings.SplitN(filter, " ", 3)
@@ -157,7 +164,14 @@ func ParseSCIMFilter(filter string) (attr, op, value string, err error) {
 
 	attr = parts[0]
 	op = strings.ToLower(parts[1])
-	value = strings.Trim(parts[2], "\"")
+
+	// Handle quoted values (may contain spaces)
+	raw := parts[2]
+	if strings.HasPrefix(raw, "\"") && strings.HasSuffix(raw, "\"") && len(raw) >= 2 {
+		value = raw[1 : len(raw)-1]
+	} else {
+		value = strings.Trim(raw, "\"")
+	}
 
 	switch op {
 	case "eq", "co", "sw":

